Authentication/SignUp/View/SignupController.swift

```diff
@@ -65,62 +67,120 @@ private extension SignupController {
     }
     
     private func bindNameElements() {
         if let nameTextField = signupView.usernameTextField {
             _viewModel.name <~ nameTextField.rex_textSignal
             _viewModel.nameValidationErrors.signal.observeNext { [unowned self] errors in
-                self._delegate.signupController(self, didFailNameValidationWithErrors: errors)
+                if errors.isEmpty {
+                    self._delegate.signupControllerDidPassNameValidation(self)
+                } else {
+                    self._delegate.signupController(self, didFailNameValidationWithErrors: errors)
+                }
             }
             if let nameValidationMessageLabel = signupView.usernameValidationMessageLabel {
                 nameValidationMessageLabel.rex_text <~ _viewModel.nameValidationErrors.signal.map { $0.first ?? " " }
             }
+            nameTextField.delegate = self
         }
     }
     
     private func bindEmailElements() {
         _viewModel.email <~ signupView.emailTextField.rex_textSignal
         _viewModel.emailValidationErrors.signal.observeNext { [unowned self] errors in
-            self._delegate.signupController(self, didFailEmailValidationWithErrors: errors)
+            if errors.isEmpty {
+                self._delegate.signupControllerDidPassEmailValidation(self)
+            } else {
+                self._delegate.signupController(self, didFailEmailValidationWithErrors: errors)
+            }
         }
         if let emailValidationMessageLabel = signupView.emailValidationMessageLabel {
             emailValidationMessageLabel.rex_text <~ _viewModel.emailValidationErrors.signal.map { $0.first ?? " " }
         }
+        signupView.emailTextField.delegate = self
     }
     
     private func bindPasswordElements() {
-        _viewModel.password <~ signupView.passwordTextField.rex_textSignal
+        _viewModel.password <~ signupView.passwordTextField.rex_textSignal.on(next: { [unowned self] text in
+            if text.isEmpty {
+                self.signupView.passwordVisibilityButton?.hidden = true
+            } else {
+                self.signupView.passwordVisibilityButton?.hidden = false
+            }
+        })
         _viewModel.passwordValidationErrors.signal.observeNext { [unowned self] errors in
-            self._delegate.signupController(self, didFailPasswordValidationWithErrors: errors)
+            if errors.isEmpty {
+                self._delegate.signupControllerDidPassPasswordValidation(self)
+            } else {
+                self._delegate.signupController(self, didFailPasswordValidationWithErrors: errors)
+            }
         }
         if let passwordValidationMessageLabel = signupView.passwordValidationMessageLabel {
             passwordValidationMessageLabel.rex_text <~ _viewModel.passwordValidationErrors.signal.map { $0.first ?? " " }
```

Comment:
@@ -6,14 +6,14 @@
  //  Copyright © 2016 Wolox. All rights reserved.		  //  Copyright © 2016 Wolox. All rights reserved.
  //		  //
  		  
 -import Foundation		
  import ReactiveCocoa		  import ReactiveCocoa
  		  
  public final class SignupController: UIViewController {		  public final class SignupController: UIViewController {
      		      
 -    private let _viewModel: SignupViewModelType		 +    private var _viewModel: SignupViewModelType
@FranDepascuali     Wolox member
FranDepascuali added a note 10 minutes ago
Why is this var?
 Add a line note
      private let _signupViewFactory: () -> SignupViewType		      private let _signupViewFactory: () -> SignupViewType
      private let _delegate: SignupControllerDelegate		      private let _delegate: SignupControllerDelegate
 +    private let _transitionDelegate: SignupControllerTransitionDelegate
      		      
      public lazy var signupView: SignupViewType = self._signupViewFactory()		      public lazy var signupView: SignupViewType = self._signupViewFactory()
      		      
 @@ -27,10 +27,11 @@ public final class SignupController: UIViewController {
      // you should not override the `createSignupController` method, but all the others 		      // you should not override the `createSignupController` method, but all the others 
      // that provide the elements this controller uses. (That is to say,		      // that provide the elements this controller uses. (That is to say,
      // `createSignupView`, `createSignupViewModel`, `createSignupControllerDelegate`)		      // `createSignupView`, `createSignupViewModel`, `createSignupControllerDelegate`)
 -    internal init(viewModel: SignupViewModelType, signupViewFactory: () -> SignupViewType, delegate: SignupControllerDelegate = DefaultSignupControllerDelegate()) {		 +    internal init(configuration: SignupControllerConfiguration) {
 -        _viewModel = viewModel		 +        _viewModel = configuration.viewModel
 -        _signupViewFactory = signupViewFactory		 +        _signupViewFactory = configuration.viewFactory
 -        _delegate = delegate		 +        _delegate = configuration.delegate
 +        _transitionDelegate = configuration.transitionDelegate
          super.init(nibName: nil, bundle: nil)		          super.init(nibName: nil, bundle: nil)
          addKeyboardObservers()		          addKeyboardObservers()
      }		      }
 @@ -49,6 +50,7 @@ public final class SignupController: UIViewController {
      		      
      		      
      public override func viewDidLoad() {		      public override func viewDidLoad() {
 +        super.viewDidLoad()
          signupView.render()		          signupView.render()
          bindViewModel()		          bindViewModel()
          let tapRecognizer = UITapGestureRecognizer(target: self, action: #selector(dismissKeyboard))		          let tapRecognizer = UITapGestureRecognizer(target: self, action: #selector(dismissKeyboard))
 @@ -65,62 +67,120 @@ private extension SignupController {
          bindPasswordElements()		          bindPasswordElements()
          bindButtons()		          bindButtons()
          setTextfieldOrder()		          setTextfieldOrder()
 +        
 +        _viewModel.signUpExecuting.observeNext { [unowned self] executing in
 +            if executing {
@FranDepascuali     Wolox member
FranDepascuali added a note 9 minutes ago
can't we do

executing ? self._delegate.signupControllerWillExecuteSignUp(self) : self._delegate.signupControllerDidExecuteSignUp(self)
```?
 Add a line note
 +                self._delegate.signupControllerWillExecuteSignUp(self)
 +            } else {
 +                self._delegate.signupControllerDidExecuteSignUp(self)
 +            }
 +            self.signupView.signUpButtonPressed = executing
 +        }
 +        
 +        _viewModel.signUpErrors.observeNext { [unowned self] in self._delegate.signupController(self, didSignUpWithError: $0) }
      }		      }
      		      
      private func bindNameElements() {		      private func bindNameElements() {
          if let nameTextField = signupView.usernameTextField {		          if let nameTextField = signupView.usernameTextField {
              _viewModel.name <~ nameTextField.rex_textSignal		              _viewModel.name <~ nameTextField.rex_textSignal
              _viewModel.nameValidationErrors.signal.observeNext { [unowned self] errors in		              _viewModel.nameValidationErrors.signal.observeNext { [unowned self] errors in
 -                self._delegate.signupController(self, didFailNameValidationWithErrors: errors)		 +                if errors.isEmpty {
 +                    self._delegate.signupControllerDidPassNameValidation(self)
@FranDepascuali     Wolox member
FranDepascuali added a note 7 minutes ago
I think we should change this long naming (brought from objc conventions) and have it written as didPassNameValidation(self)
 Add a line note
 +                } else {
 +                    self._delegate.signupController(self, didFailNameValidationWithErrors: errors)
@FranDepascuali     Wolox member
FranDepascuali added a note 6 minutes ago
I think we should change this long naming (brought from objc conventions) and have it written as didFailNameValidation(with: ).
 Add a line note
 +                }
              }		              }
              if let nameValidationMessageLabel = signupView.usernameValidationMessageLabel {		              if let nameValidationMessageLabel = signupView.usernameValidationMessageLabel {
                  nameValidationMessageLabel.rex_text <~ _viewModel.nameValidationErrors.signal.map { $0.first ?? " " }		                  nameValidationMessageLabel.rex_text <~ _viewModel.nameValidationErrors.signal.map { $0.first ?? " " }
@FranDepascuali     Wolox member
FranDepascuali added a note 4 minutes ago
I think we should express this in a different way, as it is (<~ under the hood do this) observing a new signal every time nameValidationErrors emit a value (and we try to avoid it)
 Add a line note
              }		              }
 +            nameTextField.delegate = self
          }		          }
      }		      }
      		      
      private func bindEmailElements() {		      private func bindEmailElements() {
          _viewModel.email <~ signupView.emailTextField.rex_textSignal		          _viewModel.email <~ signupView.emailTextField.rex_textSignal
          _viewModel.emailValidationErrors.signal.observeNext { [unowned self] errors in		          _viewModel.emailValidationErrors.signal.observeNext { [unowned self] errors in
 -            self._delegate.signupController(self, didFailEmailValidationWithErrors: errors)		 +            if errors.isEmpty {
 +                self._delegate.signupControllerDidPassEmailValidation(self)
@FranDepascuali     Wolox member
FranDepascuali added a note 3 minutes ago
I think we should change this long naming (brought from objc conventions) and have it written as didPassEmailValidation(self).
 Add a line note
 +            } else {
 +                self._delegate.signupController(self, didFailEmailValidationWithErrors: errors)
@FranDepascuali     Wolox member
FranDepascuali added a note 2 minutes ago
I think we should change this long naming (brought from objc conventions) and have it written as didFailEmailValidation(with: ).
 Add a line note
 +            }
          }		          }
          if let emailValidationMessageLabel = signupView.emailValidationMessageLabel {		          if let emailValidationMessageLabel = signupView.emailValidationMessageLabel {
              emailValidationMessageLabel.rex_text <~ _viewModel.emailValidationErrors.signal.map { $0.first ?? " " }		              emailValidationMessageLabel.rex_text <~ _viewModel.emailValidationErrors.signal.map { $0.first ?? " " }
@FranDepascuali     Wolox member
FranDepascuali added a note 2 minutes ago
I think we should express this in a different way, as it is (<~ under the hood do this) observing a new signal every time nameValidationErrors emit a value (and we try to avoid it)